File: android/src/main/kotlin/app/mylekha/webcontent_converter/ConversionQueue.kt
package app.mylekha.webcontent_converter

// FIFO job queue with a single busy slot, serializing every
// contentToImage, contentToPDF, and printPreview request so at most one
// runs at a time. Direct port of StartOrQueue/OnRequestFinished from the
// Windows plugin (windows/webcontent_converter_plugin.cpp) -- see that
// file's comments for why a busy caller queues instead of failing
// outright, and why the queue is capped rather than unbounded.
class ConversionQueue(private val maxQueuedRequests: Int = 32) {
    private var requestInFlight = false
    private val pendingJobs = ArrayDeque<() -> Unit>()

    // Backstop against unbounded queue growth from a caller stuck in a
    // loop. Normal bursts of calls stay well under this and just
    // queue/succeed instead of erroring.
    fun isQueueFull(): Boolean = pendingJobs.size >= maxQueuedRequests

    // Starts `job` now if the queue is idle, or queues it to run once the
    // current job (and everything already queued ahead of it) finishes,
    // in FIFO order (see onRequestFinished). Callers are expected to have
    // already checked isQueueFull() and rejected the request themselves
    // in that case -- this always runs or queues.
    fun startOrQueue(job: () -> Unit) {
        if (requestInFlight) {
            pendingJobs.addLast(job)
            return
        }
        requestInFlight = true
        job()
    }

    // Frees the busy slot and starts the next queued job, if any.
    fun onRequestFinished() {
        requestInFlight = false
        val next = pendingJobs.removeFirstOrNull() ?: return
        requestInFlight = true
        next()
    }
}

File: darwin/Classes/ConversionQueue.swift
import Foundation

/// A FIFO job queue with one busy slot, so requests serialize instead of racing.
/// Direct port of Android's `ConversionQueue` / Windows' equivalent.
/// Shared by iOS and macOS in the same way every other file in `darwin/Classes/` is.
public class ConversionQueue {
    private let maxQueuedRequests: Int
    private var pendingJobs: [() -> Void] = []
    public private(set) var requestInFlight: Bool = false

    public init(maxQueuedRequests: Int) {
        self.maxQueuedRequests = maxQueuedRequests
    }

    public func isQueueFull() -> Bool {
        return pendingJobs.count >= maxQueuedRequests
    }

    public func startOrQueue(_ job: @escaping () -> Void) {
        if !requestInFlight {
            requestInFlight = true
            job()
        } else {
            pendingJobs.append(job)
        }
    }

    public func onRequestFinished() {
        requestInFlight = false
        guard !pendingJobs.isEmpty else { return }
        let nextJob = pendingJobs.removeFirst()
        requestInFlight = true
        nextJob()
    }
}